File: java/gov/nasa/gsfc/gmsec/api5/field/BinaryField.java
/*
 * Copyright 2007-2024 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */


/**
 * @file BinaryField.java
 */

package gov.nasa.gsfc.gmsec.api5.field;

import gov.nasa.gsfc.gmsec.api5.GmsecException;
import gov.nasa.gsfc.gmsec.api5.jni.field.JNIBinaryField;



/**
 * Specialized Field class that can be used to retain raw data.
 */
public class BinaryField extends Field
{
	//! @cond
	/**
	 * @hidden
	 * This constructor is for internal GMSEC API use only.
	 * @param field A JNIBinaryField object.
	 */
	public BinaryField(JNIBinaryField field)
	{
		setInternal(field);
	}
	//! @endcond


	/**
	 * Constructor.
	 *
	 * @param name Name of the field.
	 * @param data Raw data to associate with the field.
	 *
	 * @throws IllegalArgumentException Thrown if the field name string is null, is an empty string, or is not compliant.
	 * @throws IllegalArgumentException Thrown if the byte array is null.
	 */
	public BinaryField(String name, byte[] data) throws IllegalArgumentException
	{
		this(name, data, false);
	}


	/**
	 * Constructor.
	 *
	 * @param name Name of the field.
	 * @param data Raw data to associate with the field.
	 * @param isHeader used to indicate if Field is a header field.
	 *
	 * @throws IllegalArgumentException Thrown if the field name string is null, is an empty string, or is not compliant.
	 * @throws IllegalArgumentException Thrown if the byte array is null.
	 */
	public BinaryField(String name, byte[] data, boolean isHeader) throws IllegalArgumentException
	{
		validateName(name);

		if (data == null)
		{
			throw new IllegalArgumentException("Byte array is null");
		}

		setInternal(new JNIBinaryField(name, data, isHeader));
	}


	/**
	 * Copy constructor.
	 *
	 * @param other The other BinaryField object to copy.
	 *
	 * @throws IllegalArgumentException Thrown if the given BinaryField object is null.
	 */
	public BinaryField(BinaryField other) throws IllegalArgumentException
	{
		if (other == null)
		{
			throw new IllegalArgumentException("BinaryField object is null");
		}

		setInternal(new JNIBinaryField((JNIBinaryField) Field.getInternal(other)));
	}


	/**
	 * Returns the data stored within the BinaryField.
	 *
	 * @return A byte array.
	 */
	public byte[] getValue()
	{
		JNIBinaryField jBinField = (JNIBinaryField) getInternal();

		return jBinField.getValue();
	}
}

File: framework/src/gmsec5/c/field/binary_field.cpp
/*
 * Copyright 2007-2024 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */


/** @file binary_field.cpp
 *
 *  @brief This file contains functions for the management of a BinaryField object.
 */


#include <gmsec5/c/field/binary_field.h>

#include <gmsec5/field/BinaryField.h>

#include <gmsec5/GmsecException.h>
#include <gmsec5/Status.h>


using namespace gmsec::api5;


GMSEC_Field CALL_TYPE binaryFieldCreate(const char* fieldName, const GMSEC_U8* data, size_t dataLength, GMSEC_BOOL isHeader, GMSEC_Status status)
{
	BinaryField* field = NULL;
	Status       result;

	try
	{
		field = new BinaryField(fieldName, data, dataLength, isHeader == GMSEC_BOOL::GMSEC_TRUE);
	}
	catch (const GmsecException& e)
	{
		result = Status(e);
	}

	if (status)
	{
		*(reinterpret_cast<Status*>(status)) = result;
	}

	return field;
}


GMSEC_U8 const* CALL_TYPE binaryFieldGetValue(GMSEC_Field field, GMSEC_Status status)
{
	Field*          f        = reinterpret_cast<Field*>(field);
	BinaryField*    binField = dynamic_cast<BinaryField*>(f);
	GMSEC_U8 const* data     = NULL;
	Status          result;

	if (binField)
	{
		data = binField->getValue();
	}
	else
	{
		result = Status(FIELD_ERROR, INVALID_FIELD, "Field object is not a BinaryField");
	}

	if (status)
	{
		*(reinterpret_cast<Status*>(status)) = result;
	}

	return data;
}


size_t CALL_TYPE binaryFieldGetLength(GMSEC_Field field, GMSEC_Status status)
{
	Field*       f        = reinterpret_cast<Field*>(field);
	BinaryField* binField = dynamic_cast<BinaryField*>(f);
	size_t       dataLen  = 0;
	Status       result;

	if (binField)
	{
		dataLen = binField->getLength();
	}
	else
	{
		result = Status(FIELD_ERROR, INVALID_FIELD, "Field object is not a BinaryField");
	}

	if (status)
	{
		*(reinterpret_cast<Status*>(status)) = result;
	}

	return dataLen;
}
